File: scala/src/main/scala/euler/problem/Problem22.scala
package euler.problem

object Problem22 {

  lazy val answer: Int =
    io.Source.fromFile("../problems/22-data.txt")
      .getLines.mkString.toUpperCase.split(",")
      .map(_.trim)
      .map(unquote)
      .sorted
      .map(name => name.map(_ - 'A' + 1).sum)
      .zip(Stream.from(1))
      .map({ case (x, y) => x * y })
      .sum

  val q = "\""

  def unquote(s: String): String =
    s.stripPrefix("\"").stripSuffix("\"")

}

File: haskell/src/Euler/Problem/Problem22.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE OverloadedStrings #-}

module Euler.Problem.Problem22 (answer) where

import Data.Char          ( ord )
import Data.FileEmbed     ( embedFile )
import Data.List          ( sort )
import Data.Text          ( Text )
import Data.Text.Encoding ( decodeUtf8 )

import qualified Data.Text as Text

answer :: Integer
answer = sum scores where
    names = sort $ parseNames inputText
    scores = zipWith (*) [1..] $ map wordScore names

-- "COLIN, which is worth 3 + 15 + 12 + 9 + 14 = 53"
wordScore :: Text -> Integer
wordScore = sum . (map $ fromIntegral . letterScore) . Text.unpack

-- A = 1, B = 2, etc.
letterScore :: Char -> Int
letterScore c = (ord c) - (ord 'A') + 1

parseNames :: Text -> [Text]
parseNames = (map parseName) . (Text.splitOn ",") . Text.toUpper where
    parseName = unquote . Text.strip
    unquote = (Text.dropWhile isQ) . (Text.dropWhileEnd isQ)
    isQ = (== '"')

inputText :: Text
inputText = decodeUtf8 $(embedFile "../problems/22-data.txt")
